In [2]:
# import libraries 

In [3]:
import pandas as pd
import statsmodels.api as sm

In [4]:
my_df = pd.read_csv("C:/Users/DEEKSHA/Downloads/PGPDM/Mod_7_UC module 3_Overview of ML and AI/Assignment 2/boston.csv")

In [5]:
my_df.head()

CRIM  INDUS    NOX     RM        AGE     DIS  TAX         PT  \
0  0.00632   2.31  0.538  6.575  65.199997  4.0900  296  15.300000   
1  0.02731   7.07  0.469  6.421  78.900002  4.9671  242  17.799999   
2  0.02729   7.07  0.469  7.185  61.099998  4.9671  242  17.799999   
3  0.03237   2.18  0.458  6.998  45.799999  6.0622  222  18.700001   
4  0.06905   2.18  0.458  7.147  54.200001  6.0622  222  18.700001   

            B         MV  
0  396.899994  24.000000  
1  396.899994  21.600000  
2  392.829987  34.700001  
3  394.630005  33.400002  
4  396.899994  36.200001

In [6]:
# Use Multiple Linear Regression to build your model, with the median value of owner-occupied homes as the target variable and the rest as predictors.

In [7]:
my_df.shape

(506, 10)

We have 506 rows and 10 columns in our dataset. Since this is a fairly small dataset, it is better to use stats model approach than machine learning approach.

Now let's divide our dataset into training and test sets

In [8]:
from sklearn.model_selection import train_test_split

In [9]:
x_train, x_test, y_train, y_test = train_test_split(my_df.drop('MV', axis=1), my_df['MV'], test_size=0.25, random_state=20)

Now lets add the contant term, Bo

In [10]:
x_train = sm.add_constant(x_train)

C:\Users\DEEKSHA\Anaconda3\lib\site-packages\numpy\core\fromnumeric.py:2389: FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
  return ptp(axis=axis, out=out, **kwargs)


In [11]:
x_test = sm.add_constant(x_test)

Apply OLS Method

In [12]:
my_model=sm.OLS(y_train, x_train)

We will now try and fit our model

In [13]:
result=my_model.fit()

In [14]:
print(result.summary())

                            OLS Regression Results                            
Dep. Variable:                     MV   R-squared:                       0.674
Model:                            OLS   Adj. R-squared:                  0.666
Method:                 Least Squares   F-statistic:                     84.85
Date:                Mon, 15 Jun 2020   Prob (F-statistic):           2.66e-84
Time:                        10:15:11   Log-Likelihood:                -1168.3
No. Observations:                 379   AIC:                             2357.
Df Residuals:                     369   BIC:                             2396.
Df Model:                           9                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         12.9228      6.048      2.137      0.0

In the above summary, we observe that the p-values for predictors: Tax, CRIM and INDUS are greater than 0.05 at a Confidence Interval of 5%. This implies that these predictors do not significantly impact our target variable. We will now drop these from out dataset

In [15]:
x_train= x_train.drop({'TAX','CRIM','INDUS'}, axis=1)


In [16]:
print(x_train)

     const    NOX     RM         AGE      DIS         PT           B
283    1.0  0.401  7.923   24.799999   5.8850  13.600000  395.519989
341    1.0  0.442  7.241   49.299999   7.0379  15.500000  394.739990
107    1.0  0.520  6.127   85.199997   2.1224  20.900000  387.690002
454    1.0  0.713  6.728   94.099998   2.4961  20.200001    6.680000
353    1.0  0.410  6.728   36.099998  12.1265  17.000000  384.459992
..     ...    ...    ...         ...      ...        ...         ...
412    1.0  0.597  4.628  100.000000   1.5539  20.200001   28.790001
223    1.0  0.507  6.618   80.800003   3.2721  17.400000  396.899994
271    1.0  0.464  6.240   16.299999   4.4290  18.600000  396.899994
474    1.0  0.584  5.427   95.400002   2.4298  20.200001  352.579987
355    1.0  0.413  5.936   19.500000  10.5857  22.000000  376.040008

[379 rows x 7 columns]


In [17]:
# Now apply our model once again and check the results

In [18]:
updated_model_results=sm.OLS(y_train, x_train).fit()

In [19]:
print(updated_model_results.summary())

                            OLS Regression Results                            
Dep. Variable:                     MV   R-squared:                       0.672
Model:                            OLS   Adj. R-squared:                  0.666
Method:                 Least Squares   F-statistic:                     126.9
Date:                Mon, 15 Jun 2020   Prob (F-statistic):           8.36e-87
Time:                        10:15:11   Log-Likelihood:                -1169.8
No. Observations:                 379   AIC:                             2354.
Df Residuals:                     372   BIC:                             2381.
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         12.8361      5.921      2.168      0.0

In [20]:
from sklearn.metrics import r2_score

In [21]:
x_test=x_test.drop({'TAX','CRIM','INDUS'}, axis=1)


In [22]:
predictions = updated_model_results.predict(sm.add_constant(x_test))

In [24]:
pd.DataFrame({'Actual_value': y_test, 'Predicted_value': predictions}).sample(5)

Actual_value  Predicted_value
49      19.400000        18.929175
254     21.900000        22.814274
362     20.799999        10.496821
181     36.200001        26.123404
406     11.900000         5.119348

In [49]:
r2_score(y_test, predictions)

0.5827534947564261

Although we have removed the features that were not significantly affecting our MV values, the r2_score is still not close to 1. Hence it is difficult to establish a good acuracy of our model. One reason for such result could be use of raw and unclean data here. No prior eda or data cleaning has been done and data used as is. Another reason could be high multicollinearity among the other features as stated above in the warnings.

Let us try to check other parameters to check our model's accuracy

In [50]:
updated_model_results.rsquared_adj

0.6664294676841895

In [51]:
updated_model_results.rsquared

0.6717242380384087

In [52]:
updated_model_results.pvalues

const    3.080545e-02
NOX      2.563689e-07
RM       1.536955e-47
AGE      2.502242e-02
DIS      2.769692e-07
PT       5.168696e-14
B        1.633848e-05
dtype: float64

Estimating coefficients and intercept

In [53]:
updated_model_results.params

const    12.836079
NOX     -23.046613
RM        7.143374
AGE      -0.036487
DIS      -1.186605
PT       -1.106702
B         0.014353
dtype: float64

In [54]:
updated_model_results.conf_int()

0          1
const   1.192913  24.479246
NOX   -31.678677 -14.414548
RM      6.307331   7.979417
AGE    -0.068372  -0.004602
DIS    -1.632345  -0.740864
PT     -1.384660  -0.828744
B       0.007890   0.020815

In [57]:
# from metrics class in scikit-learn, import more error metrics

In [62]:
from sklearn.metrics import mean_squared_error, mean_absolute_error

# Mean Absolute Error - 
MAE of your model refers to the mean of the absolute values of each prediction error on all instances of the test data-set. Prediction error is the difference between the actual value and the predicted value for that instance.

In [28]:
# MAE 
mean_absolute_error(y_test, predictions)

3.8456340500493504

In [29]:
# MSE 
mean_squared_error(y_test, predictions)

33.63236400618135